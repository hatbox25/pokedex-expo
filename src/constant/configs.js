export const BASE_API_URL = "https://pokeapi.co/api/v2/";
export const API_TIMEOUT = 15000; //milisecond
