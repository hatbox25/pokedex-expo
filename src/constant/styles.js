export default {
  primaryColor: "rgb(246,207,83)",
  secondaryColor: "rgb(24,26,31)",
  backgroundColor: "rgb(234,234,234)",
};
