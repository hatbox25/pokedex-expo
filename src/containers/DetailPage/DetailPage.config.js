import PropTypes from "prop-types";

const displayName = "DetailPage";
const propTypes = {
  pokemon: PropTypes.shape({
    data: PropTypes.object,
    loading: PropTypes.loading,
  }),
  getPokemon: PropTypes.func.isRequired,
  route: PropTypes.object,
};
const defaultProps = {};

export default { displayName, propTypes, defaultProps };
