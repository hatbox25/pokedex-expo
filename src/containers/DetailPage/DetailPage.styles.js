import defaultStyle from "../../constant/styles";

export default {
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },

  content: {
    alignItems: "center",
  },

  imageContainer: {
    flexDirection: "row",
    width: "100%",
    backgroundColor: defaultStyle.secondaryColor,
    padding: 16,
    justifyContent: "space-around",
  },
  image: {
    height: 120,
    width: 120,
  },

  detailContainer: {
    padding: 16,
    paddingBottom: 0,
    width: "100%",
  },

  listContainer: {
    width: "100%",
    backgroundColor: defaultStyle.secondaryColor,
    padding: 16,
    alignItems: "center",
  },
  listTitle: {
    color: defaultStyle.primaryColor,
    fontWeight: "600",
    marginBottom: 16,
  },
  listItem: { color: defaultStyle.backgroundColor, textAlign: "center" },

  statContainer: {
    width: "100%",
    padding: 16,
    alignItems: "center",
  },
  statTitle: {
    color: defaultStyle.secondaryColor,
    fontWeight: "600",
    marginBottom: 16,
  },
};
