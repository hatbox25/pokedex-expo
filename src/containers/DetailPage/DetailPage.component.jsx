import React, { PureComponent } from "react";
import { View, ScrollView, Text, ActivityIndicator, Image } from "react-native";
import get from "lodash.get";
import { capitalizeWord } from "../../utils/string.util";
import TextValue from "../../components/TextValue/TextValue.component";

import config from "./DetailPage.config";
import styles from "./DetailPage.styles";

import defaultStyle from "../../constant/styles";

class DetailPage extends PureComponent {
  componentDidMount() {
    const {
      getPokemon,
      route: {
        params: { pokemonId }, // get pokemon id pushed by navigation
      },
    } = this.props;
    getPokemon(pokemonId).then((result) => {
      this.props.navigation.setOptions({
        title: `Detail Pokemon: ${capitalizeWord(result.name)}`,
      });
    });
  }

  renderPokemonImages = (data) => (
    <View style={styles.imageContainer}>
      <Image
        style={styles.image}
        source={{
          uri: get(data, "sprites.front_default"),
        }}
      />
      <Image
        style={styles.image}
        source={{
          uri: get(data, "sprites.back_default"),
        }}
      />
    </View>
  );

  renderGeneralInfo = (data) => (
    <View style={styles.detailContainer}>
      <TextValue label="Base Experience" value={get(data, "base_experience")} />

      <TextValue label="Height" value={get(data, "height")} />

      <TextValue label="Weight" value={get(data, "weight")} />
    </View>
  );

  renderListInfo = (title, list, key) => (
    <View style={styles.listContainer}>
      <Text style={styles.listTitle}>{title}</Text>
      <Text style={styles.listItem}>
        {list.map((item, index) =>
          index === 0
            ? get(item, `${key}.name`)
            : `, ${get(item, `${key}.name`)}`
        )}
      </Text>
    </View>
  );

  renderStats = (data) => (
    <View style={styles.statContainer}>
      <Text style={styles.statTitle}>Stats</Text>
      {get(data, "stats", []).map((stat) => (
        <TextValue
          label={capitalizeWord(get(stat, "stat.name"))}
          value={get(stat, "base_stat")}
        />
      ))}
    </View>
  );

  render() {
    const { data, loading } = this.props.pokemon;
    if (loading)
      return (
        <View style={styles.container}>
          <ActivityIndicator size="large" color={defaultStyle.primaryColor} />
        </View>
      );
    return (
      <ScrollView contentContainerStyle={styles.content}>
        {this.renderPokemonImages(data)}
        {this.renderGeneralInfo(data)}
        {this.renderListInfo("Types", get(data, "types", []), "type")}
        {this.renderStats(data)}
        {this.renderListInfo(
          "Abilities",
          get(data, "abilities", []),
          "ability"
        )}
        {this.renderListInfo("Moves", get(data, "moves", []), "move")}
      </ScrollView>
    );
  }
}

DetailPage.displayName = config.displayName;
DetailPage.defaultProps = config.defaultProps;
DetailPage.propTypes = config.propTypes;

export default DetailPage;
