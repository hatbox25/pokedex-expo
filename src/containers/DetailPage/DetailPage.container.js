import { connect } from "react-redux";

import DetailPage from "./DetailPage.component";

import { getPokemon } from "../../redux/actions/detail.action";

const mapStateToProps = (state) => ({
  pokemon: state.detail,
});
const mapDispatchToProps = { getPokemon };

export default connect(mapStateToProps, mapDispatchToProps)(DetailPage);
