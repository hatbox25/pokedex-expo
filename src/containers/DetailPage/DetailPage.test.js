import React from "react";
import { shallow } from "enzyme";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

import DetailPageContainer from "./DetailPage.container";
import DetailPage from "./DetailPage.component";
import config from "./DetailPage.config";

import { detailPokemon } from "../../constant/fictures";

const mockStore = configureStore([thunk]);
const initialState = {};
const store = mockStore(initialState);

const props = {
  ...config.defaultProps,
  pokemon: {
    data: detailPokemon,
    loading: false,
  },
  route: {
    params: {
      pokemonId: "1",
    },
  },
  getPokemon: jest.fn(() => Promise.resolve()),
};

let wrapper;
beforeEach(() => {
  wrapper = shallow(<DetailPage {...props} />);
});

afterEach(() => {
  jest.clearAllMocks();
});

describe("DetailPage Snap Test", () => {
  test("should render correctly on finished get", () => {
    expect(shallow(<DetailPageContainer store={store} />)).toMatchSnapshot();
    expect(wrapper).toMatchSnapshot();
  });

  test("should render correctly on loading", () => {
    wrapper.setProps({
      pokemon: {
        data: detailPokemon,
        loading: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
