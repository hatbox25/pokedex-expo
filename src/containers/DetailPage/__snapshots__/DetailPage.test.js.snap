// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailPage Snap Test should render correctly on finished get 1`] = `
<ContextProvider
  value={null}
>
  <DetailPage
    getPokemon={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[`DetailPage Snap Test should render correctly on finished get 2`] = `
<ScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgb(24,26,31)",
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
        }
      }
      style={
        Object {
          "height": 120,
          "width": 120,
        }
      }
    />
    <Image
      source={
        Object {
          "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png",
        }
      }
      style={
        Object {
          "height": 120,
          "width": 120,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "padding": 16,
        "paddingBottom": 0,
        "width": "100%",
      }
    }
  >
    <TextValue
      label="Base Experience"
      value={64}
    />
    <TextValue
      label="Height"
      value={7}
    />
    <TextValue
      label="Weight"
      value={69}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(24,26,31)",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(246,207,83)",
          "fontWeight": "600",
          "marginBottom": 16,
        }
      }
    >
      Types
    </Text>
    <Text
      style={
        Object {
          "color": "rgb(234,234,234)",
          "textAlign": "center",
        }
      }
    >
      grass
      , poison
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(24,26,31)",
          "fontWeight": "600",
          "marginBottom": 16,
        }
      }
    >
      Stats
    </Text>
    <TextValue
      key="hp"
      label="Hp"
      value={45}
    />
    <TextValue
      key="attack"
      label="Attack"
      value={49}
    />
    <TextValue
      key="defense"
      label="Defense"
      value={49}
    />
    <TextValue
      key="special-attack"
      label="Special-attack"
      value={65}
    />
    <TextValue
      key="special-defense"
      label="Special-defense"
      value={65}
    />
    <TextValue
      key="speed"
      label="Speed"
      value={45}
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(24,26,31)",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(246,207,83)",
          "fontWeight": "600",
          "marginBottom": 16,
        }
      }
    >
      Abilities
    </Text>
    <Text
      style={
        Object {
          "color": "rgb(234,234,234)",
          "textAlign": "center",
        }
      }
    >
      overgrow
      , chlorophyll
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgb(24,26,31)",
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "rgb(246,207,83)",
          "fontWeight": "600",
          "marginBottom": 16,
        }
      }
    >
      Moves
    </Text>
    <Text
      style={
        Object {
          "color": "rgb(234,234,234)",
          "textAlign": "center",
        }
      }
    >
      razor-wind
      , swords-dance
      , cut
      , bind
      , vine-whip
      , headbutt
      , tackle
      , body-slam
      , take-down
      , double-edge
      , growl
      , strength
      , mega-drain
      , leech-seed
      , growth
      , razor-leaf
      , solar-beam
      , poison-powder
      , sleep-powder
      , petal-dance
      , string-shot
      , toxic
      , rage
      , mimic
      , double-team
      , defense-curl
      , light-screen
      , reflect
      , bide
      , sludge
      , skull-bash
      , amnesia
      , flash
      , rest
      , substitute
      , snore
      , curse
      , protect
      , sludge-bomb
      , mud-slap
      , giga-drain
      , endure
      , charm
      , swagger
      , fury-cutter
      , attract
      , sleep-talk
      , return
      , frustration
      , safeguard
      , sweet-scent
      , synthesis
      , hidden-power
      , sunny-day
      , rock-smash
      , facade
      , nature-power
      , ingrain
      , knock-off
      , secret-power
      , grass-whistle
      , bullet-seed
      , magical-leaf
      , natural-gift
      , worry-seed
      , seed-bomb
      , energy-ball
      , leaf-storm
      , power-whip
      , captivate
      , grass-knot
      , venoshock
      , round
      , echoed-voice
      , grass-pledge
      , work-up
      , grassy-terrain
      , confide
    </Text>
  </View>
</ScrollView>
`;

exports[`DetailPage Snap Test should render correctly on loading 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="rgb(246,207,83)"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
